internal/configs/provider_validation.go

```diff
@@ -95,6 +142,44 @@ func validateProviderConfigs(parentCall *ModuleCall, cfg *Config, noProviderConf
 		}
 	}
 
+	checkImpliedProviderNames := func(resourceConfigs map[string]*Resource) {
+		// Now that we have all the provider configs and requirements validated,
+		// check for any resources which use an implied localname which doesn't
+		// match that of required_providers
+		for _, r := range resourceConfigs {
+			// We're looking for resources with no specific provider reference
+			if r.ProviderConfigRef != nil {
+				continue
+			}
+
+			localName := r.Addr().ImpliedProvider()
+			if _, ok := localNames[localName]; ok {
+				// OK, this was listed directly in the required_providers
+				continue
+			}
+
+			defAddr := addrs.NewDefaultProvider(localName)
+
+			// Now make sure we don't have the same provider required under a
+			// different name.
+			for prevLocalName, addr := range localNames {
+				if addr.Equals(defAddr) {
+					diags = append(diags, &hcl.Diagnostic{
+						Severity: hcl.DiagWarning,
+						Summary:  "Duplicate required provider",
+						Detail: fmt.Sprintf(
+							"Provider %q was implicitly required via resource %q, but listed in required_providers as %q. Either the local name in required_providers must match the resource name, or the %q provider must be assigned within the resource block.",
+							localName, r.Addr(), prevLocalName, prevLocalName,
```

Comment:
Subjectively, I think it'd help a little to make that first `%q` be `defAddr` instead of just `localName`, even though _we_ know that it'll always just be `hashicorp/` followed by the `localName` anyway, just to add a little nudge here for those who aren't yet aware of this implied dependency mechanism and might benefit from us linking it with what they can see in the `terraform init` and `terraform providers` output.

Probably we'll refine this message some more once we see what sorts of questions people ask about it, since it's hard to guess right now what a typical recipient of this message might know or not know, so otherwise I think this is a fine place to start!

Reply:
I actually had that at first, but changed it as I cut the message down for brevity ;) THat's a fair point though, and will put it back.